File: hs/src/Libertree/Model/CommentLike.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-warn-unused-do-bind #-}

module Libertree.Model.CommentLike where

import Libertree.Model.Common

import qualified Database.PostgreSQL.Simple as Db
import Data.Int (Int64)
import Data.Aeson (encode, object, (.=))
import Data.Text (Text)

data CommentLikeM = CommentLikeM
                      Int64         -- member id
                      Int64         -- comment id
                      (Maybe Int64) -- remote id

instance Model CommentLikeM where
  create db entity@(CommentLikeM m c r) =
    Db.withTransaction db $ do
      Db.execute db
        "INSERT INTO comment_likes (member_id, comment_id, remote_id) VALUES (?, ?, ?)"
          (m, c, r)
      distribute db entity

instance Distributable CommentLikeM where
  isLocal (CommentLikeM _ _ Nothing) = True
  isLocal _ = False
  distribute db e =
    if isLocal e
      then do
        [x] :: [Db.Only Int64] <- Db.query_ db "SELECT currval('comment_likes_id_seq')"
        enqueue db
          "distribution"
          "Request::COMMENT-LIKE"
          (encode $ object [ "comment_like_id" .= Db.fromOnly x ])
      else return 0

File: lib/libertree/model/comment-like.rb
module Libertree
  module Model
    class CommentLike < Sequel::Model(:comment_likes)
      def after_create
        super
        if self.local? && self.comment.post.distribute?
          Libertree::Model::Job.create_for_forests(
            {
              task: 'request:COMMENT-LIKE',
              params: { 'comment_like_id' => self.id, }
            },
            *self.forests
          )
        end
      end

      def local?
        self.remote_id.nil?
      end

      def member
        @member ||= Member[self.member_id]
      end

      def comment
        @comment ||= Comment[self.comment_id]
      end

      def before_destroy
        if self.local? && self.comment.post.distribute?
          Libertree::Model::Job.create_for_forests(
            {
              task: 'request:COMMENT-LIKE-DELETE',
              params: { 'comment_like_id' => self.id, }
            },
            *self.forests
          )
        end
        super
      end

      # TODO: the correct method to call is "destroy"
      def delete
        self.before_destroy
        super
      end

      def delete_cascade
        self.before_destroy
        DB.dbh[ "SELECT delete_cascade_comment_like(?)", self.id ].get
      end

      def self.create(*args)
        like = super
        like.comment.notify_about_like like
        like
      end

      def forests
        if self.comment.post.remote?
          self.comment.post.server.forests
        else
          Libertree::Model::Forest.all_local_is_member
        end
      end
    end
  end
end
